skafka/src/main/scala/com/evolutiongaming/skafka/consumer/RebalanceCallback.scala

```diff
@@ -0,0 +1,333 @@
+
+  def seek(partition: TopicPartition, offsetAndMetadata: OffsetAndMetadata): RebalanceCallback[Nothing, Unit] =
+    WithConsumer(_.seek(partition.asJava, offsetAndMetadata.asJava))
+
+  def seekToBeginning(partitions: Nes[TopicPartition]): RebalanceCallback[Nothing, Unit] =
+    WithConsumer(_.seekToBeginning(partitions.asJava))
+
+  def seekToEnd(partitions: Nes[TopicPartition]): RebalanceCallback[Nothing, Unit] =
+    WithConsumer(_.seekToEnd(partitions.asJava))
+
+  def subscription: RebalanceCallback[Nothing, Set[Topic]] = WithConsumer(_.subscription().asScala.toSet)
+
+  private[consumer] def run[F[_]: ToTry, A](
+    rebalanceCallback: RebalanceCallback[F, A],
+    consumer: ConsumerJ[_, _]
+  ): Try[A] = {
+    type S = Any => RebalanceCallback[F, Any]
+
+    @tailrec
+    def loop[A1](c: RebalanceCallback[F, A1], ss: List[S]): Try[Any] = {
+      c match {
+        case c: Pure[A1] =>
+          ss match {
+            case Nil     => c.a.pure[Try]
+            case s :: ss => loop(s(c.a), ss)
+          }
+        case c: Lift[F, A1] =>
+          c.fa.toTry match {
+            case Success(a) =>
+              ss match {
+                case Nil     => a.pure[Try]
+                case s :: ss => loop(s(a), ss)
+              }
+            case Failure(a) => a.raiseError[Try, A1]
+          }
+        case c: WithConsumer[A1] =>
+          Try { c.f(consumer) } match {
+            case Success(a) =>
+              ss match {
+                case Nil     => a.pure[Try]
+                case s :: ss => loop(s(a), ss)
+              }
+            case Failure(a) => a.raiseError[Try, A1]
+          }
+        case c: Bind[F, _, A1] =>
+          loop(c.source, c.fs.asInstanceOf[S] :: ss)
+        case Error(a) =>
+          a.raiseError[Try, A1]
+      }
+    }
+
+    for {
+      a <- loop(rebalanceCallback, List.empty)
+      a <- Try { a.asInstanceOf[A] }
+    } yield a
+  }
+
+  private final case class Pure[+A](a: A) extends RebalanceCallback[Nothing, A]
+
+  private final case class Bind[F[_], S, +A](source: RebalanceCallback[F, S], fs: S => RebalanceCallback[F, A])
+      extends RebalanceCallback[F, A]
+
+  private final case class Lift[F[_], A](fa: F[A]) extends RebalanceCallback[F, A]
+
+  private final case class WithConsumer[+A](f: ConsumerJ[_, _] => A) extends RebalanceCallback[Nothing, A]
+
+  private final case class Error(throwable: Throwable) extends RebalanceCallback[Nothing, Nothing]
+
+  implicit class RebalanceCallbackOps[F[_], A](val self: RebalanceCallback[F, A]) extends AnyVal {
+
+    def map[B](f: A => B): RebalanceCallback[F, B] = flatMap(a => pure(f(a)))
+
+    def flatMap[B](f: A => RebalanceCallback[F, B]): RebalanceCallback[F, B] = Bind(self, f)
+
+    def mapK[G[_]](fg: F ~> G): RebalanceCallback[G, A] = {
+      self match {
+        case Pure(a)          => Pure(a)
+        case Bind(source, f)  => Bind(source.mapK(fg), f andThen (_.mapK(fg)))
```

Comment:
interesting :)

I believe this may cause stackoverflow for huge enough data structure, however I don't think you should fix this.

Reply:
created https://github.com/evolution-gaming/skafka/issues/130 just in case